action.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+use dokuwiki\plugin\twofactor\Provider;
+use dokuwiki\Form\Form;
+
+// Load the PHP QR Code library.
+require_once(dirname(__FILE__).'/phpqrcode.php');
+
+/**
+ * If we turn this into a helper class, it can have its own language and settings files.
+ * Until then, we can only use per-user settings.
+ */
+class action_plugin_twofactorgoogleauth extends Provider
+{
+
+    /**
+     * @inheritDoc
+     */
+    public function isConfigured()
+    {
+        return $this->getConf('enable') === 1 &&
+            $this->settings->get('secret') &&
+            $this->settings->get('verified');
+    }
+
+    /**
+     * This user will need to interact with the QR code in order to configure GA.
+     */
+    public function renderProfileForm(Form $form)
+    {
+        global $conf;
+        global $USERINFO;
+
+        if ($this->settings->get('secret')) { // The user has a revokable GA secret.
+            // Show the QR code so the user can add other devices.
+            $mysecret = $this->settings->get('secret');
+            $data = $this->generateQRCodeData($USERINFO['name'].'@'.$conf['title'], $mysecret);
+            $form->addHTML('<figure><figcaption>'.$this->getLang('directions').'</figcaption>');
+            $form->addHTML('<img src="'.$data.'" alt="'.$this->getLang('directions').'" />');
+            $form->addHTML('</figure>');
+            // Check to see if the user needs to verify the code.
+            if (!$this->settings->get('verified')) {
+                $form->addHTML('<span>'.$this->getLang('verifynotice').'</span>');
+                $form->addTextInput(
+                    'googleauth_verify',
+                    $this->getLang('verifymodule'),
+                    ''
+                );
+            }
+            // Show the option to revoke the GA secret.
+            $form->addCheckbox('googleauth_disable', $this->getLang('killmodule'));
```

Comment:
revocation (the deletion of all data for this module) is handled in the main class